funsor/ops/op.py

```diff
@@ -1,9 +1,30 @@
 # Copyright Contributors to the Pyro project.
 # SPDX-License-Identifier: Apache-2.0
 
+import weakref
+
 from multipledispatch import Dispatcher
 
 
+class OpCacheMeta(type):
+    """
+    Metaclass for caching op instance construction.
+    """
+    # use a WeakKeyDictionary to allow garbage collection of out-of-scope op classes.
+    # this is especially important when some ops are dynamically generated and
+    # their instances are large, e.g. an op corresponding to a neural network.
+    _cls_caches = weakref.WeakKeyDictionary()  # class -> {args: instance}
+
+    def __call__(cls, *args, **kwargs):
```

Comment:
Unless there is a very good reason to allow `**kwargs`, I would recommend removing it and forcing users to pass positional arguments.